lib/Dialect/TritonGPU/Transforms/Pipeliner/AssignLatencies.cpp

```diff
@@ -0,0 +1,259 @@
+#include "triton/Analysis/AxisInfo.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/PipeliningUtility.h"
+#include "triton/Dialect/TritonGPU/Transforms/Schedule.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+#include "triton/Dialect/TritonNvidiaGPU/IR/Dialect.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "triton-pipeline-schedule"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+using namespace mlir;
+namespace tt = mlir::triton;
+namespace ttg = mlir::triton::gpu;
+namespace ttng = mlir::triton::nvidia_gpu;
+
+namespace mlir {
+namespace triton {
+namespace gpu {
+
+namespace {
+
+// Return true if the preconditions for pipelining the loop are met.
+bool preCondition(scf::ForOp forOp) {
```

Comment:
That's my thinking too, at some point there may be different preCoditions for different pieces, so I kept them separate, just refactored the internals so it more modular.

Reply:
I think the preconditions are mostly for performance reasons and belong to the scheduling. Ideally we should be able to pipeline other cases even if they don't go through shared memory and it may not be efficient, the expander part of the pipeliner shouldn't have to make this call

Reply:
I think we may still need preCondidtions for expander, due to limitations in what expander is actually capable of pipelining (no dist-2 deps, etc). These will be likely different from preConditions for scheduling (which are perf related), so I duplicated them, because they are likely to diverge. Does it make sense?

Reply:
got it, makes sense